src/Util/Autoload.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace BetterReflection\Util;
+
+use BetterReflection\Reflection\ReflectionClass;
+use PhpParser\Node\Name;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\PrettyPrinter\Standard as CodePrinter;
+
+class Autoload
+{
+    /**
+     * @var ReflectionClass[]
+     */
+    private static $reflections = [];
+
+    /**
+     * @var bool
+     */
+    private static $initialiseCalled = false;
+
+    /**
+     * @param ReflectionClass $reflectionClass
+     * @return void
+     * @throws \RuntimeException
+     * @throws \LogicException
+     */
+    public static function addClass(ReflectionClass $reflectionClass)
+    {
+        if (!self::$initialiseCalled) {
+            // @todo specific exception
+            throw new \LogicException('You must call Autoload::initialise() before adding reflections to load');
+        }
+
+        if (array_key_exists($reflectionClass->getName(), self::$reflections)) {
+            // @todo specific exception
+            throw new \LogicException(sprintf('Class %s already registered', $reflectionClass->getName()));
+        }
+
+        if (class_exists($reflectionClass->getName(), false)) {
+            // @todo specific exception
+            throw new \RuntimeException(sprintf('Class already loaded: %s', $reflectionClass->getName()));
+        }
+
+        self::$reflections[$reflectionClass->getName()] = $reflectionClass;
+    }
+
+    /**
+     * Register our own autoloader
+     * @throws \LogicException
+     * @return void
+     */
+    public static function initialise()
+    {
+        if (self::$initialiseCalled) {
+            // @todo specific exception
+            throw new \LogicException('Autoload::initialise has already been called');
+        }
+
+        spl_autoload_register([__CLASS__, 'autoload'], true, true);
+
+        self::$initialiseCalled = true;
+    }
+
+    /**
+     * Reset our autoloader, unregistering it and returning it to a clean state
+     * @return void
+     */
+    public static function reset()
```

Comment:
Can't make `private function __invoke`  - has to be public, so it can't be callable unless we want to expose the autoloader (not sure if that's a good idea?)